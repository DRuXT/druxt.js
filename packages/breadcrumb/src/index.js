import { DruxtBreadcrumbModule } from './module'

/**
 * The Nuxt.js module function.
 *
 * Installs the module functionality in a Nuxt.js frontend.
 *
 * @type {Function}
 * @exports default
 * @name DruxtBreadcrumbModule
 * @see {@link ./module|DruxtBreadcrumbModule}
 */
export default DruxtBreadcrumbModule

/**
 * Vue.js Mixin.
 *
 * @type {object}
 * @exports DruxtBreadcrumbMixin
 * @see {@link ./mixins/breadcrumb|DruxtBreadcrumbMixin}
 */
 export { DruxtBreadcrumbMixin } from './mixins/breadcrumb'
