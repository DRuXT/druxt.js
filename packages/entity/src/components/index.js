/**
 * The `<DruxtEntity />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtEntity
 * @see {@link ./DruxtEntity|DruxtEntity}
 */
export { default as DruxtEntity } from './DruxtEntity.vue'

/**
 * The `<DruxtEntityForm />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtEntityForm
 * @see {@link ./DruxtEntityForm|DruxtEntityForm}
 */
export { default as DruxtEntityForm } from './DruxtEntityForm.vue'

/**
 * The `<DruxtEntityFormButtons />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtEntityFormButtons
 * @see {@link ./DruxtEntityFormButtons|DruxtEntityFormButtons}
 */
export { default as DruxtEntityFormButtons } from './DruxtEntityFormButtons.vue'

/**
 * The `<DruxtField />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtField
 * @see {@link ./DruxtField|DruxtField}
 */
export { default as DruxtField } from './DruxtField.vue'

/**
 * The `<DruxtRouterEntity />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtRouterEntity
 * @see {@link ./DruxtRouterEntity|DruxtRouterEntity}
 */
export { default as DruxtRouterEntity } from './DruxtRouterEntity.vue'

export * from './fields'
