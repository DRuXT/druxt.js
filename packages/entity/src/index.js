/**
 * Vue.js mixins.
 *
 * @type {object}
 * @exports DruxtEntityMixins
 * @see {@link ./mixins|DruxtEntityMixins}
 * @see {@link ./mixins/entity|DruxtEntityMixin}
 * @see {@link ./mixins/field|DruxtFieldMixin}
 */
export * from './mixins'

/**
 * Default function to alert user to incorrectly installed module.
 *
 * This was added as part of the @nuxt/kit update due to breaking changes.
 */
export default () => {
  throw new Error("DruxtEntity Nuxt module must be installed as 'druxt-entity/nuxt'")
}
