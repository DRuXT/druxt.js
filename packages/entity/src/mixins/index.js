export * from './context.js'
export * from './componentSuggestion.js'

/**
 * Vue.js mixin for DruxtEntity and DruxtEntityForm Wrapper components.
 *
 * @type {object}
 * @exports DruxtEntityMixin
 * @see {@link ./entity|DruxtEntityMixin}
 */
export { DruxtEntityMixin } from './entity'

/**
 * Vue.js mixin for DruxtField Wrapper components.
 *
 * @type {object}
 * @exports DruxtFieldMixin
 * @see {@link ./field|DruxtFieldMixin}
 */
export { DruxtFieldMixin } from './field'
