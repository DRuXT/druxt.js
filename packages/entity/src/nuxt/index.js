import { defineNuxtModule, installModule } from '@nuxt/kit'
import { join } from 'path'
// import DruxtEntityStorybook from '../nuxtStorybook'

/**
 * The Nuxt.js module function.
 *
 * - Adds dependant modules.
 * - Adds Nuxt plugin.
 * - Adds Nuxt Storybook integration.
 *
 * The module function should not be used directly, but rather installed via your Nuxt configuration file.
 *
 * Options are set on the root level `druxt` Nuxt config object.
 *
 * @example @lang js
 * // `nuxt.config.js`
 * module.exports = {
 *   modules: [
 *     'druxt-entity'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 *
 * @todo Document options.
 *
 * @property {object} options.druxt - DruxtJS root level options.
 * @property {string} options.druxt.baseUrl - Base URL of Drupal JSON:API backend.
 */
const DruxtEntityNuxtModule = defineNuxtModule({
  meta: {
    name: 'druxt-entity',
  },
  defaults: {
    baseUrl: '',
    endpoint: '/jsonapi'
  },

  async setup(moduleOptions, nuxt) {
    // Set default options.
    const options = {
      baseUrl: moduleOptions.baseUrl,
      ...nuxt.options?.druxt,
      entity: {
        query: {},
        ...nuxt.options?.druxt?.entity,
        ...moduleOptions,
        components: {
          fields: false,
          ...nuxt.options?.druxt?.entity?.components,
          ...moduleOptions?.components
        }
      }
    }

    // Add dependant modules.
    await installModule('druxt/nuxt', options, nuxt)
    await installModule('druxt-schema/nuxt', { baseUrl: options.baseUrl }, nuxt)

    // Register components directories.
    nuxt.hook('components:dirs', dirs => {
      dirs.push({
        path: join(__dirname, '../dist/components'),
        ignore: ['fields']
      })
      if (options.entity.components.fields) {
        dirs.push({ path: join(__dirname, '../dist/components/fields') })
      }
    })

    // Nuxt Storybook.
    // @TODO - @nuxt/kit and @nuxt/storybook aren't compatible.
    // this.nuxt.hook('storybook:config', async ({ stories }) => {
    //   await DruxtEntityStorybook.call(this, { stories })
    // })
  }
})

export default DruxtEntityNuxtModule
