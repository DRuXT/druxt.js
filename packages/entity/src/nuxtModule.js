import { resolve } from 'path'
import DruxtEntityStorybook from './nuxtStorybook'

/**
 * The Nuxt.js module function.
 *
 * - Adds dependant modules.
 * - Adds Nuxt plugin.
 * - Adds Nuxt Storybook integration.
 *
 * The module function should not be used directly, but rather installed via your Nuxt configuration file.
 *
 * Options are set on the root level `druxt` Nuxt config object.
 *
 * @example @lang js
 * // `nuxt.config.js`
 * module.exports = {
 *   modules: [
 *     'druxt-entity'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 *
 * @todo Add module level options.
 * @todo Document options.
 *
 * @property {object} options.druxt - DruxtJS root level options.
 * @property {string} options.druxt.baseUrl - Base URL of Drupal JSON:API backend.
 */
const DruxtEntityNuxtModule = function (moduleOptions = {}) {
  const { options } = this

  // Use root level Druxt options.
  if (typeof options === 'undefined' || !options.druxt) {
    throw new TypeError('Druxt settings missing.')
  }

  // Add dependant modules.
  const modules = ['druxt', 'druxt-schema'].filter((module) => !options.modules.includes(module))
  for (const module of modules) {
    this.addModule(module)
  }

  // Add plugin.
  this.addPlugin({
    src: resolve(__dirname, '../nuxt/plugin.js'),
    fileName: 'druxt-entity.js',
    options: options.druxt
  })

  // Nuxt Storybook.
  this.nuxt.hook('storybook:config', async ({ stories }) => {
    await DruxtEntityStorybook.call(this, { stories })
  })
}

DruxtEntityNuxtModule.meta = require('../package.json')

export { DruxtEntityNuxtModule }
