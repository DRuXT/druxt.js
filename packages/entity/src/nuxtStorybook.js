import { resolve } from 'path'
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'
import { DruxtClient } from 'druxt'

// @todo - Use component naming convention for title.
const titleFn = (parts) =>
  parts.map((part) =>
    part.charAt(0).toUpperCase() + part.slice(1).replace(/_/g, ' ')
  ).join('/')

export default async function ({ stories }) {
  const { addTemplate, options } = this

  const druxt = new DruxtClient(options.druxt.baseUrl, options.druxt)

  // Fetch and process schemas.
  const [view, form] = (await Promise.all(['view', 'form'].map(async (type) =>
    await druxt.getCollectionAll(`entity_${type}_display--entity_${type}_display`, new DrupalJsonApiParams()
      .addFilter('status', 1)
      .addFields(`entity_${type}_display--entity_${type}_display`, [
        'bundle',
        'mode',
        'targetEntityType',
      ])
    )))).map((schema) => schema.map((collection) => collection.data.map((o) => o)).flat())
  const schemas = [...view, ...form]

  // Get Entity type and bundle data.
  const entityTypes = Array.from(new Set(schemas.map((o) => o.attributes.targetEntityType)))
    .map((entity) => ({
      entity,
      bundles: Array.from(new Set(schemas.filter((o) => o.attributes.targetEntityType === entity).map((o) => o.attributes.bundle))),
    })
  )

  // Fetch Entities for UUID controls.
  const entities = await Promise.all([
    ...entityTypes.map(({ entity, bundles }) => bundles.map(async (bundle) => {
      try {
        return {
          resourceType: `${entity}--${bundle}`,
          entities: (await druxt.getCollection(`${entity}--${bundle}`, new DrupalJsonApiParams()
            .addFilter('status', 1)
            .addFields(`${entity}--${bundle}`, ['id', 'title', 'name', 'info'])
          )).data.map((o) => ({
            id: o.id,
            title: (o.attributes || {}).title || (o.attributes || {}).name || (o.attributes || {}).info || o.id,
          })),
        }
      }
      catch(e) {
        return {
          resourceType: `${entity}--${bundle}`,
          entities: [],
        }
      }
    })).flat()
  ])

  // Build Entity Stories.
  const entityStories = entityTypes.map(({ entity, bundles }) => bundles.map((bundle) => {
    return Object.entries({ view, form }).map(([type, schemas]) => {
      const displays = Array.from(new Set(schemas.filter((o) => o.attributes.targetEntityType === entity && o.attributes.bundle === bundle).map((o) => o.attributes.mode).sort((a) => (a === 'default' ? -1 : 0))))
      if (!displays) {
        return false
      }

      const resourceType = `${entity}--${bundle}`
      const component = type === 'view' ? 'druxt-entity' : 'druxt-entity-form'

      addTemplate({
        src: resolve(__dirname, `../nuxt/${component}.stories.js`),
        fileName: `stories/${component}.${resourceType}.stories.js`,
        options: {
          displays,
          entities: entities.find((o) => o.resourceType === resourceType).entities,
          resourceType,
          title: titleFn(['Druxt Entity', entity, bundle, `${type} displays`]),
        },
      })
  
      return resolve(options.buildDir, `./stories/${component}.${resourceType}.stories.js`)
    })
  })).flat(2)

  stories.push.apply(stories, entityStories)
}
