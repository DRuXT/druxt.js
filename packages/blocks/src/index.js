import { DruxtBlocksNuxtModule } from './nuxtModule'

/**
 * The Nuxt.js module function.
 *
 * Installs the module functionality in a Nuxt application.
 *
 * @type {Function}
 * @exports default
 * @name DruxtBlocksModule
 * @see {@link ./nuxtModule|DruxtBlocksModule}
 *
 * @example <caption>nuxt.config.js</caption> @lang js
 * module.exports = {
 *   modules: [
 *     'druxt-blocks'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 */
export default DruxtBlocksNuxtModule

/**
 * Vue.js Mixin.
 *
 * @type {object}
 * @exports DruxtBlocksBlockMixin
 * @see {@link ./mixins/block|DruxtBlocksBlockMixin}
 */
 export { DruxtBlocksBlockMixin } from './mixins/block'
