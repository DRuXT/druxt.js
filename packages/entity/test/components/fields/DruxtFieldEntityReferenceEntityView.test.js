import 'regenerator-runtime/runtime'
import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'
import mockAxios from 'jest-mock-axios'

import { DruxtClient, DruxtStore } from '../../../../druxt/src'
import DruxtFieldEntityReferenceEntityView from '../../../src/components/fields/DruxtFieldEntityReferenceEntityView.vue'

jest.mock('axios')

const baseURL = 'https://example.com'

// Setup local vue instance.
const localVue = createLocalVue()
localVue.use(Vuex)

const stubs = ['druxt-entity']
let store

const mountComponent = (options) => {
  const entity = {
    type: 'pages',
    id: 'fe00c55d-0335-49d6-964e-a868c0c68f9c',
    attributes: {
      title: 'Welcome to Contenta CMS!',
      path: {
        alias: '/welcome'
      }
    }
  }
  store.commit('druxt/addResource', { resource: { data: entity }})

  const propsData = {
    value: [{
      type: entity.type,
      uuid: entity.id
    }],
    schema: {}
  }

  return shallowMount(DruxtFieldEntityReferenceEntityView, { ...options, localVue, propsData, store, stubs })
}

describe('Component - DruxtFieldEntityReferenceEntityView', () => {
  beforeEach(() => {
    mockAxios.reset()

    // Setup vuex store.
    store = new Vuex.Store()

    DruxtStore({ store })
    store.$druxt = new DruxtClient('https://demo-api.druxtjs.org')

    store.app = { context: { error: jest.fn() }, store }
  })

  test('default', async () => {
    const wrapper = mountComponent()

    await localVue.nextTick()
    await localVue.nextTick()

    expect(wrapper.vm.mode).toBe('default')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
