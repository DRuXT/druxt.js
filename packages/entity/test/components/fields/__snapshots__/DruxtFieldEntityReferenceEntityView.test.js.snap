// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - DruxtFieldEntityReferenceEntityView default 1`] = `
"<div>
  <!---->
  <!---->
  <druxtentity-stub wrapper=\\"[object Object]\\" mode=\\"default\\" type=\\"pages\\" uuid=\\"fe00c55d-0335-49d6-964e-a868c0c68f9c\\"></druxtentity-stub>
</div>"
`;
